storage/fanout.go

```diff
@@ -240,27 +240,40 @@ func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher)
 func (q *mergeQuerier) SelectSorted(params *SelectParams, matchers ...*labels.Matcher) (SeriesSet, Warnings, error) {
 	seriesSets := make([]SeriesSet, 0, len(q.queriers))
 	var warnings Warnings
+
+	queryResultChan := make(chan *queryResult)
 	for _, querier := range q.queriers {
-		set, wrn, err := querier.SelectSorted(params, matchers...)
-		q.setQuerierMap[set] = querier
-		if wrn != nil {
-			warnings = append(warnings, wrn...)
-		}
-		if err != nil {
-			q.failedQueriers[querier] = struct{}{}
+		go func(qr Querier) {
+			set, wrn, err := qr.SelectSorted(params, matchers...)
+			queryResultChan <- &queryResult{qr: qr, set: set, wrn: wrn, selectError: err}
+		}(querier)
+	}
+	for i := 0; i < len(q.queriers); i++ {
+		qryResult := <-queryResultChan
+		q.setQuerierMap[qryResult.set] = qryResult.qr
+		if qryResult.wrn != nil {
+			warnings = append(warnings, qryResult.wrn...)
+		}
+		if qryResult.selectError != nil {
+			q.failedQueriers[qryResult.qr] = struct{}{}
 			// If the error source isn't the primary querier, return the error as a warning and continue.
-			if querier != q.primaryQuerier {
-				warnings = append(warnings, err)
-				continue
+			if qryResult.qr != q.primaryQuerier {
+				warnings = append(warnings, qryResult.selectError)
 			} else {
-				return nil, nil, err
+				return nil, nil, qryResult.selectError
 			}
 		}
-		seriesSets = append(seriesSets, set)
 	}
 	return NewMergeSeriesSet(seriesSets, q), warnings, nil
 }
 
+type queryResult struct {
```

Comment:
Put this inside the function, as you're only using it there.

Reply:
done. 

![image](https://user-images.githubusercontent.com/9583245/74016098-d2fa5a80-49cc-11ea-9497-288ccd9a0fa1.png)